pkg/operator/watcher/filewatcher.go

```diff
@@ -0,0 +1,137 @@
+package watcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"sync"
+
+	"gopkg.in/fsnotify.v1"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+)
+
+var log = logf.Logger.WithName("filewatcher")
+
+// FileWatcher watches a file for changes.
+type FileWatcher struct {
+	sync.Mutex
+	fileName    string
+	currentData []byte
+	watcher     *fsnotify.Watcher
+}
+
+// New returns a new FileWatcher watching the given file.
+func New(file string) (*FileWatcher, error) {
+	fw := &FileWatcher{
+		fileName: file,
+	}
+
+	// Initial read of file.
+	if err := fw.ReadFile(); err != nil {
+		return nil, err
+	}
+
+	if watcher, err := fsnotify.NewWatcher(); err != nil {
+		return nil, err
+	} else {
+		fw.watcher = watcher
+	}
+
+	return fw, nil
+}
+
+// GetFileData fetches the data of the currently watched file.
+func (fw *FileWatcher) GetFileData() []byte {
+	fw.Lock()
+	defer fw.Unlock()
+	return fw.currentData
+}
+
+// Start starts the FileWatcher.
+func (fw *FileWatcher) Start(stopCh <-chan struct{}, reloadCh chan bool) error {
+	if err := fw.watcher.Add(fw.fileName); err != nil {
+		return err
+	}
+
+	go fw.Watch(reloadCh)
+
+	log.Info("starting file watcher")
+
+	// Block until the stop channel is closed.
+	<-stopCh
+
+	return fw.watcher.Close()
+}
+
+// Watch reads events from the watcher's channel and reacts to changes.
+func (fw *FileWatcher) Watch(reload chan bool) {
+	for {
+		select {
+		case event, ok := <-fw.watcher.Events:
+			// Channel is closed.
+			if !ok {
+				return
```

Comment:
Probably should log that the events channel was closed before returning (ditto for the error channel below).

Reply:
done